### PR TITLE
[968] Add deferred status badge and date to trainee review page

### PR DESCRIPTION
**Pull request by avinhurry**
### Context

Adding a new `trainee status` row to the review page. This page will have a dynamic status depending on the state of the trainee. If the trainee is in a `withdrawn` or `deferred` state, the date this occurred will be shown too. 

### Changes proposed in this pull request

- Refactor trainee-status class
- Add trainee status row to review page
- Add dynamic status tag
- Add date of deferral and withdrawal for these trainees
- Add view components for trainee status row
- Add view components for all the different types of status tags

### Guidance to review

- Pull down the branch or visit https://rtt-review-pr-471.herokuapp.com/
- Choose a withdrawn trainee and add a date to `trainee.withdraw_date` in rails c
- Check that the date and status tags are visible in the trainee status row on the review page
- Repeat for a deferred trainee
- Check that the `trainee status row` does not render if the trainee is in any other state
- Inspect the view components added in this PR 
- 🚢 

## Screenshots of the trainee status row

image

image

### CODE REVIEW

**Comment by avinhurry** [comment]
[Add defered status badge to trainee review page](https://trello.com/c/LjiSc8fM/968-add-defered-status-badge-to-trainee-review-page)

File: spec/components/trainees/record_details/view_spec.rb
Comment (gpeng) [inline spec/components/trainees/record_details/view_spec.rb:46]: Should this be a `let`?

File: app/components/trainees/record_details/view.rb
Comment (AbigailMcP) [inline app/components/trainees/record_details/view.rb:55]: I would prefer to return early in situations like this, if there's no `status_date` then you also don't need to go ahead and calculate the `status_date_prefix`:

```suggestion
      def trainee_status_date
        status_date = trainee.deferred? ? trainee.defer_date : trainee.withdraw_date
        return unless status_date
        
        status_date_prefix = trainee.deferred? ? "Deferral date: " : "Withdrawal date: "
        status_date_prefix + date_for_summary_view(status_date)
      end
```

File: app/components/trainees/record_details/view.rb
Comment (AbigailMcP) [inline app/components/trainees/record_details/view.rb:45]: Does this mean there's an extra `<br>` for all trainee statuses that don't have a date?

File: spec/components/trainees/record_details/view_spec.rb
Comment (avinhurry) [reply spec/components/trainees/record_details/view_spec.rb]: Thanks boss, change made.

**Review by AbigailMcP — COMMENTED** [review]
Nice! Couple of points below.

This also raises a point about our example data - we shouldn't really have withdrawn or deferred trainees without the respective dates as this isn't possible in the UI.

I'll record in a separate ticket.

File: app/components/trainees/record_details/view.rb
Comment (avinhurry) [reply app/components/trainees/record_details/view.rb:55]: Thanks! I was trying to figure out the best way to write this for ages.

File: app/components/trainees/record_details/view.rb
Comment (avinhurry) [reply app/components/trainees/record_details/view.rb:45]: Yes, is there something else that you would suggest?

File: app/components/trainees/record_details/view.html.erb
Comment (AbigailMcP) [inline app/components/trainees/record_details/view.html.erb:11]: As discussed, this should only be present if the trainee is deferred or withdrawn.

File: app/components/trainees/record_details/view.rb
Comment (AbigailMcP) [reply app/components/trainees/record_details/view.rb:45]: As discussed this won't be an issue once we're only rendering the row for withdrawn and deferred trainees - there'll _always_ be a date.

**Comment by ndgallon73** [comment]
Hi, looks good from a status perspective, both defer & withdrawal appear as expected. One quick question before I approve though is that every time you change state (i.e click on the defer action) in this version you get an error page, see attachment. On returning to the records list though the correct status is applied. Is this a bug that has been introduced?

**Comment by avinhurry** [comment]
> Hi, looks good from a status perspective, both defer & withdrawal appear as expected. One quick question before I approve though is that every time you change state (i.e click on the defer action) in this version you get an error page, see attachment. On returning to the records list though the correct status is applied. Is this a bug that has been introduced?

That wasn't introduced with this PR. I've seen it before. I think it's something to do with connecting to DTTP. Defer to @gpeng for more info 🤷🏼‍♂️

File: app/components/status_tag/view.rb
Comment (AbigailMcP) [inline app/components/status_tag/view.rb:4]: We have a standard pattern of supplying extra classes to view components - see `FormComponents::Autocomplete::View ` as an example. You should be able to add your `"trainee_status"` to `default_classes` and call `super` in the `initialize`.

**Review by ndgallon73 — APPROVED** [review]
Approved from a product perspective.

File: app/components/status_tag/view.rb
Comment (avinhurry) [reply app/components/status_tag/view.rb]: Thanks for your help with this!

**Review by AbigailMcP — APPROVED** [review]

**Pull request merged**
